gobblin-core/src/main/java/gobblin/async/AsyncDataDispatcher.java

```diff
@@ -0,0 +1,149 @@
+package gobblin.async;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.Queue;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+
+/**
+ * Base class with skeleton logic to dispatch a record asynchronously. It buffers the records and consumes
+ * them by {@link RecordProcessor}
+ *
+ * <p>
+ *   However the records are consumed depends on the actual implementation of {@link #dispatch(Queue)}, which
+ *   may process one record or a batch at a time
+ * </p>
+ *
+ * @param <D> type of record
+ */
+@ThreadSafe
+public abstract class AsyncDataDispatcher<D> implements Closeable {
```

Comment:
This seems like an `AbstractIdleService` rather than a `Closeable`, which also provides the `ProcessorState` for free.